src/lib/insights/listener.tsx

```diff
@@ -9,23 +9,37 @@ type WithInsightsListenerProps = {
   insights: InsightsClientWrapper;
 };
 
+const findRealTarget = (
+  startElement: HTMLElement | null,
+  endElement: HTMLElement | null
+): HTMLElement | null => {
+  let element: HTMLElement | null = startElement;
+  while (element && !hasDataAttributes(element)) {
+    if (element === endElement) {
+      return null;
+    }
+    element = element.parentElement;
+  }
+  return element;
+};
+
 const insightsListener = (BaseComponent: any) => {
   function WithInsightsListener(props: WithInsightsListenerProps) {
     const handleClick = (event: MouseEvent): void => {
-      if (!hasDataAttributes(event.target as HTMLElement)) {
-        return;
-      }
-
       if (!props.insights) {
         throw new Error(
           'The `insightsClient` option has not been provided to `instantsearch`.'
         );
       }
-
-      const { method, payload } = readDataAttributes(
-        event.target as HTMLElement
+      const realTarget = findRealTarget(
```

Comment:
A better naming for this variable is welcome 😄

Reply:
`insightsRoot` seems fine.

Reply:
what would `root` mean here?

Reply:
What `findRealTarget` (current naming) is doing here is resolving the root HTML element where data attributes are defined. The root is where the data needs to be read from.

`insightsRootTarget` if you want to keep the "target" name.

Reply:
Ok I understand, but actually the code we're adding here is trying to find a DOM elements (not necessarily a root element) that carries the data-insights attribute, so I'm not sure how `root` adds meaning here.
Is `insightsTarget` fine or would it be something missing?

Reply:
They're both fine. Root just means it's the highest parent in this context.

Reply:
ok, renamed to `insightsTarget` in 115c932f